### ISSUE TITLE
map + filter with unwrap on options and results

### ISSUE THREAD

**Issue by nyurik** [rust-lang/rust-clippy#15026]
### What it does

Given a sequence of `Option` types like `&[Option<T>]`, users sometimes tend to iterate with `filter + map(unwrap)` sequence. Instead, they should use `map_filter` without the unwrap.

Other cases:
* sequence of `Result`s
* usage of other "unwrapping" functions like `unwrap_or_default`, `expect`, etc

### Advantage

- Faster and cleaner code without panic

### Example

```rust
pub fn filter_and_map_options(values: &[Option<&i32>]) -> String {
    values
        .iter()
        .filter(|s| s.is_some())
        .map(|s| s.unwrap().to_string())
        .collect::<String>()
}
```

Could be written as:

```rust
pub fn filter_and_map_options(values: &[Option<&i32>]) -> String {
    values
        .iter()
        .filter_map(|s| s.map(|v| v.to_string()))
        .collect::<String>()
}
```

**Comment by matthiaskrgr** [comment]
could also use `flatten()` I think?

```rust
pub fn filter_and_map_options(values: &[Option<&i32>]) -> String {
    values
        .iter()
        .flatten()
        .map(|x| x.to_string())
        .collect::<String>()
}
```

**Comment by nyurik** [comment]
oh, good idea, thanks!  I didn't even realize this.  Also, I wounder if we should have more lints like these?

* `iter.filter(|v| v.is_some()).map(...)` - always suggest `iter.flatten()`, even if we cannot detect if there is an `unwrap` - simply because the chances of user wanting to preserve a sequence of `Option`s are low.  Sadly, this is less than certain.
* same, but if `map` clearly unwraps the value - high confidence `iter.flatten()`, with the result probably compiling ok
* `iter.filter_map(|v| v.map(|v| ...))` - treat it the same as above
